[Excel] (Custom functions) Localization (#859)
OfficeDev/office-js-docs-pr: docs/excel/custom-functions-localize.md

> ---

Reviewer: Don't forget to add this article to the TOC!

ms.date: 03/15/2019
description: Localize your Excel custom functions. 
title: Localization of Excel custom functions (preview)
localization_priority: Normal
---
# Localize custom functions

> Just like add-ins, custom functions can be localized. You'll provide locale information in the XML manifest file of your add-in and provide your own localized function names in the functions' JSON file.

Reviewer: ```suggestion
To make your custom functions work around the world, localize them into different languages. To localize custom functions, you'll need to provide locale information in the XML manifest file and provide localized function names in the functions' JSON file.
```

Good to lead with the benefit (also custom functions are add-ins...)

## Localize your add-in

> Before localizing your function names, you'll need to add optional locales to your XML file. For more information on this process, see [Localization for Office Add-ins](../develop/localization.md#control-localization-from-the-manifest).

Reviewer: Would it be possible to show the locale for the German example below for manifest.xml?

Author: Yep

## Localize function names

> To declare a localized function name, you will set alternative `name` and `description` properties within your function's JSON file. You can declare multiple localized names and descriptions.

Reviewer: ```suggestion
To declare a localized function name, create additional `name` and `description` properties within your function's JSON file. You can declare multiple localized names and descriptions.
```

> Note your functions' `name` and `description` can change by locale, the `id` of each function is not localized. This `id` property is how Excel identifies your function as unique and should not be changed once it is set.

Reviewer: ```suggestion
The `name` and `description` appear in Excel and are localized. However, the `id` of each function is not localized. The `id` property is how Excel identifies your function as unique and should not be changed once it is set.
```

> In the following code sample, you'll see the JSON file for a function with the `id` property "multiply." The `name` and `description` property of the function has been localized for German. Note that the parameter `names` and `descriptions` have also been localized for German.

Reviewer: ```suggestion
In the following code sample, you'll see the JSON file for a function with the `id` property "MULTIPLY." The `name` and `description` property of the function is localized for German. Each parameter `name` and `description` is also localized for German.
```

> 

Reviewer: Would be great if this also showed the English version to stress the point that you create multiple sections for each language.

Author: Ok can do!

```JSON
{
    "id": "MULTIPLY",
    "name": "SUMME",
    "description": "Summe zwei Zahlen",
    "helpUrl": "http://www.contoso.com",
    "result": {
        "type": "number",
        "dimensionality": "scalar"
    },
    "parameters": [
        {
            "name": "eins",
            "description": "Erste Nummer",
            "dimensionality": "scalar"
        },
        {
            "name": "zwei",
            "description": "Zweite Nummer",
            "dimensionality": "scalar"
        },
    ],
}
```

Compare the previous code sample with the English version that follows.

```JSON
{
    "id": "MULTIPLY",
    "name": "Multiply",
    "description": "Multiplies two numbers",
    "helpUrl": "http://www.contoso.com",
    "result": {
        "type": "number",
        "dimensionality": "scalar"
    },
    "parameters": [
        {
            "name": "one",
            "description": "first number",
            "dimensionality": "scalar"
        },
        {
            "name": "two",
            "description": "second number",
            "dimensionality": "scalar"
        },
    ],
}
```

## Localize your add-in

> In addition to providing two JSON files, you need to provide your XML manifest file with an override value for the URL of your functions' script file. The following code sample shows a default `en-us` locale with an override functions script URL for `de-de` (Germany). The override file is a separate script file with localized German function names and ids.

Reviewer: ```suggestion
After creating a JSON file for each language, you need to update your XML manifest file with an override value for each locale that specifies the URL of each JSON metadata file. The following manifest XML shows a default `en-us` locale with an override JSON file URL for `de-de` (Germany). The **functions-de.json** file contains the localized German function names and ids.
```

```XML
<DefaultLocale>en-us</DefaultLocale>
...
<Resources>
     <bt:Urls>

>         <bt:Url id="Contoso.Functions.Script.Url" DefaultValue="https://localhost:3000/dist/functions.js"/>

Reviewer: ```suggestion
        <bt:Url id="Contoso.Functions.Metadata.Url" DefaultValue="https://localhost:3000/dist/functions.json"/>
```

>           <bt:Override Locale="de-de" Value="https://localhost:3000/src/functions/functionsInGerman.js" />

Reviewer: ```suggestion
          <bt:Override Locale="de-de" Value="https://localhost:3000/dist/functions-de.json" />
        </bt:url>
        
```

     </bt:Urls>
</Resources>
```

> If you're using the Yo Office generator to make your custom functions project, make sure your webpack.config.js file is properly configured to handle the proper script files.

Reviewer: ```suggestion
If you're using the Yo Office generator to make your custom functions project, make sure your webpack.config.js file is properly configured to handle additional JSON files.
```

---

> ms.date: 04/25/2019

Author: ```suggestion
ms.date: 04/30/2019
```

description: Localize your Excel custom functions. 
title: Localize custom functions (preview)
localization_priority: Normal
---

# Localize custom functions

> To make your custom functions work around the world, localize them into different languages. To localize custom functions, you need to provide localized function names in the functions' JSON file and provide locale information in the XML manifest file. Autogenerated metadata does not work for localization so you need to update the JSON file manually.

Author: ```suggestion
To make your custom functions work around the world, localize them into different languages. To localize custom functions, you need to provide localized function names in the functions' JSON file and provide locale information in the XML manifest file. Autogenerated metadata doesn't work for localization so you need to update the JSON file manually.
```

    "helpUrl": "http://www.contoso.com",
    "result": {
        "type": "number",
        "dimensionality": "scalar"
    },
    "parameters": [
        {
            "name": "one",
            "description": "first number",
            "dimensionality": "scalar"
        },
        {
            "name": "two",
            "description": "second number",
            "dimensionality": "scalar"
        },
    ],
}
```

## Localize your add-in

After creating a JSON file for each language, you need to update your XML manifest file with an override value for each locale that specifies the URL of each JSON metadata file. The following manifest XML shows a default `en-us` locale with an override JSON file URL for `de-de` (Germany). The **functions-de.json** file contains the localized German function names and ids.

```XML
<DefaultLocale>en-us</DefaultLocale>
...
<Resources>
     <bt:Urls>
        <bt:Url id="Contoso.Functions.Metadata.Url" DefaultValue="https://localhost:3000/dist/functions.json"/>
          <bt:Override Locale="de-de" Value="https://localhost:3000/dist/functions-de.json" />
        </bt:url>
        
     </bt:Urls>
</Resources>
```

> If you're using the Yo Office generator to make your custom functions project, make sure your webpack.config.js file is properly configured to handle additional JSON files.

Author: ```suggestion
```
Actually this works out ok if everything is being handled already through dist
